### PR TITLE
Hermes 

### PR DESCRIPTION
**Pull request by gabrielleitezup**
Hermes folder structure
https://app.clubhouse.io/charle/story/42805/cria%C3%A7%C3%A3o-da-estrutura-do-projeto-hermes

### CODE REVIEW

File: hermes/resources/hermes/cometarios.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+- Controle de threads para o consumo da fila
```
Comment (barbararochazup) [inline hermes/resources/hermes/cometarios.txt:1]: acho que seria legal remover esse arquivo depois

File: hermes/resources/hermes/scripts.sql [hunk 2]
```diff
@@ -0,0 +1,62 @@
+-- Hermes
```
Comment (barbararochazup) [inline hermes/resources/hermes/scripts.sql:1]: precisa manter esse file? a mesma estrutura está no arquivo inicial do migration

**Review by barbararochazup — CHANGES_REQUESTED** [review]
algumas duvidas e sugestões

File: hermes/internal/configuration/configuration.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+)
+
+var initialValues = map[string]string{
+	"DB_USER":        "charlescd_hermes",
+	"DB_PASSWORD":    "charlescd_hermes",
+	"DB_HOST":        "localhost",
+	"DB_NAME":        "charlescd_hermes",
+	"DB_SSL":         "disable",
+	"DB_PORT":        "5432",
+	"ENCRYPTION_KEY": "maycon",
```
Comment (arthurpuccizup) [inline hermes/internal/configuration/configuration.go:39]: put a real key here

File: hermes/internal/subscription/subscription.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+			WithOperations("Delete.DeleteSubscription")
+	}
+
+	return nil
+}
+
+func (main Main) FindById(subscriptionId uuid.UUID) (Response, errors.Error) {
+	result := Response{}
+
+	subsQuery := main.db.Raw(FindOneQuery(subscriptionId.String())).Row()
```
Comment (arthurpuccizup) [inline hermes/internal/subscription/subscription.go:131]: Verify if this is injection safe, the query is built using string placeholders with

File: hermes/internal/subscription/queries.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+		"CreatedBy": createdBy,
+	}
+
+}
+
+func FindOneQuery(subscriptionId string) string {
+	return fmt.Sprintf(`SELECT external_id, url, description, PGP_SYM_DECRYPT(api_key, '%s'), events
+	FROM SUBSCRIPTIONS
+	WHERE  id = '%s'
+	AND deleted_at IS NULL`, configuration.GetConfiguration("ENCRYPTION_KEY"), subscriptionId)
```
Comment (arthurpuccizup) [inline hermes/internal/subscription/queries.go:52]: is this placeholder substitution injection safe?

File: hermes/internal/configuration/configuration.go (hunk 3 above)
Comment (arthurpuccizup) [inline hermes/internal/configuration/configuration.go:39]: What about use only the .env file and when the variable is not found we could throw an error, or at startup check if all variables have values

File: hermes/web/api/middleware.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	return ""
+}
+
+func ValidatorMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-type", "application/json")
+		workspaceID := r.Header.Get("x-workspace-id")
+		ers := util.NewApiErrors()
+
+		fmt.Println(r.RequestURI)
```
Comment (arthurpuccizup) [inline hermes/web/api/middleware.go:52]: what the purpose of this print call?

File: hermes/web/api/serverapi.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	return router
+}
+
+func (api *Api) health(r *mux.Router) {
+	r.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(":)"))
+		return
+	})
+}
+
```
Comment (arthurpuccizup) [inline hermes/web/api/serverapi.go:61]: i think it's a good ideia we have a readiness api too, with the check if the database and the queue are both up

File: hermes/web/api/subscription/subscription.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+ */
+
+package subscription
+
+import (
+	"errors"
+	"github.com/google/uuid"
+	"github.com/gorilla/mux"
+	"hermes/internal/subscription"
+	util2 "hermes/web/util"
```
Comment (arthurpuccizup) [inline hermes/web/api/subscription/subscription.go:26]: why its util2? there no other util import here, if it's necessary, why not call it restUtil?

File: hermes/web/util/error.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+package util
+
+import "hermes/pkg/errors"
+
+type ErrorLink map[string]string
+
+type ApiError struct {
+	errors.SimpleError
```
Comment (arthurpuccizup) [inline hermes/web/util/error.go:26]: i'm worried about returning some internal error specificity in this, maybe we can generalize the returns to return a generic message or could return a specific one but with no technical information in it, like database information

File: hermes/web/api/subscription/subscription.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+		params := mux.Vars(r)
+		subscriptionId, uuidErr := uuid.Parse(params["subscriptionId"])
+		if uuidErr != nil {
+			util2.NewResponse(w, http.StatusInternalServerError, uuidErr)
+			return
+		}
+
+		createdSubscription, err := subscriptionMain.Update(subscriptionId, request)
+		if err != nil {
+			util2.NewResponse(w, http.StatusInternalServerError, err)
```
Comment (arthurpuccizup) [inline hermes/web/api/subscription/subscription.go:77]: the error stuff problems

File: hermes/web/api/subscription/subscription.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+		author := r.Header.Get("x-author")
+		if author == "" {
+			util2.NewResponse(w, http.StatusInternalServerError, errors.New("author is required"))
+			return
+		}
+
+		err := subscriptionMain.Delete(subscriptionId, author)
+		if err != nil {
+			util2.NewResponse(w, http.StatusInternalServerError, err)
```
Comment (arthurpuccizup) [inline hermes/web/api/subscription/subscription.go:102]: the error stuff problems

File: hermes/web/api/subscription/subscription.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+		params := mux.Vars(r)
+		subscriptionId, uuidErr := uuid.Parse(params["subscriptionId"])
+		if uuidErr != nil {
+			util2.NewResponse(w, http.StatusInternalServerError, uuidErr)
+			return
+		}
+
+		result, err := subscriptionMain.FindById(subscriptionId)
+		if err != nil {
+			util2.NewResponse(w, http.StatusInternalServerError, err)
```
Comment (arthurpuccizup) [inline hermes/web/api/subscription/subscription.go:121]: the error stuff problems

File: hermes/internal/messageexecutionhistory/main.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+package messageexecutionhistory
+
+import (
+	"gorm.io/gorm"
+	"hermes/pkg/errors"
+)
+
+type UseCases interface {
+	Save(execution Request) (Response, errors.Error)
+}
```
Comment (arthurpuccizup) [inline hermes/internal/messageexecutionhistory/main.go:28]: No unit tests

File: hermes/internal/message/main.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"gorm.io/gorm"
+	"hermes/pkg/errors"
+	"io"
+)
+
+type UseCases interface {
+	ParseMessage(subscription io.ReadCloser) (Request, errors.Error)
+	Save(message Request) (ExecutionResponse, errors.Error)
+}
```
Comment (arthurpuccizup) [inline hermes/internal/message/main.go:30]: No unit tests

File: hermes/internal/subscription/main.go [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+type UseCases interface {
+	ParseSubscription(subscription io.ReadCloser) (Request, errors.Error)
+	ParseUpdate(subscription io.ReadCloser) ([]byte, errors.Error)
+	Validate(subscription Request) errors.ErrorList
+	Save(subscription Request) (SaveResponse, errors.Error)
+	Update(subscriptionId uuid.UUID, subscription []byte) (UpdateResponse, errors.Error)
+	Delete(subscriptionId uuid.UUID, author string) errors.Error
+	FindById(subscriptionId uuid.UUID) (Response, errors.Error)
+}
```
Comment (arthurpuccizup) [inline hermes/internal/subscription/main.go:36]: No unit tests

**Review by arthurpuccizup — CHANGES_REQUESTED** [review]

File: hermes/go.mod [hunk 16]
```diff
@@ -0,0 +1,19 @@
+module hermes
+
+go 1.14
```
[the remark this answers is not in the document]
Comment (gabrielleitezup) [reply hermes/go.mod]: Go version updated

File: hermes/internal/subscription/queries.go (hunk 5 above)
Comment (gabrielleitezup) [reply hermes/internal/subscription/queries.go]: Now it is. (Query removed :D)

File: hermes/internal/subscription/subscription.go (hunk 4 above)
Comment (gabrielleitezup) [reply hermes/internal/subscription/subscription.go]: Solved

File: hermes/web/api/middleware.go (hunk 6 above)
Comment (gabrielleitezup) [reply hermes/web/api/middleware.go]: removed

File: hermes/internal/configuration/configuration.go (hunk 3 above)
Comment (gabrielleitezup) [reply hermes/internal/configuration/configuration.go]: done

Comment (gabrielleitezup) [reply hermes/internal/configuration/configuration.go]: Done (with sadness..)

File: hermes/web/api/serverapi.go (hunk 7 above)
Comment (gabrielleitezup) [reply hermes/web/api/serverapi.go:71]: Done

File: hermes/web/api/subscription/subscription.go (hunk 8 above)
Comment (gabrielleitezup) [reply hermes/web/api/subscription/subscription.go]: Done

File: hermes/web/util/error.go (hunk 9 above)
Comment (gabrielleitezup) [reply hermes/web/util/error.go]: Will be fixed in a future PR.

File: hermes/web/api/subscription/subscription.go (hunk 10 above)
Comment (gabrielleitezup) [reply hermes/web/api/subscription/subscription.go]: Will be fixed in a future PR.

Comment (gabrielleitezup) [reply hermes/web/api/subscription/subscription.go]: Will be fixed in a future PR.

File: hermes/web/api/subscription/subscription.go (hunk 11 above)
Comment (gabrielleitezup) [reply hermes/web/api/subscription/subscription.go]: Will be fixed in a future PR.

File: hermes/web/api/subscription/subscription.go (hunk 12 above)
Comment (gabrielleitezup) [reply hermes/web/api/subscription/subscription.go]: Will be fixed in a future PR.

File: hermes/internal/messageexecutionhistory/main.go (hunk 13 above)
Comment (gabrielleitezup) [reply hermes/internal/messageexecutionhistory/main.go:28]: Will be fixed in a future PR.

File: hermes/internal/message/main.go (hunk 14 above)
Comment (gabrielleitezup) [reply hermes/internal/message/main.go:30]: Will be fixed in a future PR.

File: hermes/internal/subscription/main.go (hunk 15 above)
Comment (gabrielleitezup) [reply hermes/internal/subscription/main.go]: Will be fixed in a future PR.

**Pull request closed without merging**